### DISCUSSION TITLE
WCF serviceThrottling on CoreWCF - is it possible

### DISCUSSION THREAD

**Discussion by pacojones** [CoreWCF/CoreWCF#1008] [General]
Hi all,
in .Net framework, I was able to set some throttling parameters, on serviceBehavior, like this:

`<behaviors>
      <serviceBehaviors>
        <behavior name="MyServiceBehavior">
          <serviceDebug includeExceptionDetailInFaults="true" />
          <serviceMetadata httpGetEnabled="false" httpsGetEnabled="true" />
          <serviceSecurityAudit auditLogLocation="Application"
            suppressAuditFailure="true"
            serviceAuthorizationAuditLevel="Failure"
            messageAuthenticationAuditLevel="Failure" />
          <serviceThrottling maxConcurrentCalls="500" maxConcurrentInstances="500" maxConcurrentSessions="500"/>
          <dataContractSerializer maxItemsInObjectGraph="2147483647" />
        </behavior>
      </serviceBehaviors>
    </behaviors>`

Is there any way of achieving this on CoreWCF?

**Comment by navichandan** [comment]
We are also looking for the same, any help on this would be appreciated.

**Comment by iyrwis** [comment]
Also have this need. I go through the CoreWCF source-code and not know how to do it.

**Comment by Kuling** [comment]
Hi everybody
I have similar need. I need limit MaxConcurrentCalls for my service.
Maybe some of you find my research useful.

--

While investigating these issues https://github.com/CoreWCF/CoreWCF/issues/857, https://github.com/CoreWCF/CoreWCF/issues/547 I found in source code that there is class [ServiceThrottle](https://github.com/CoreWCF/CoreWCF/blob/main/src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/ServiceThrottle.cs) which is suitable for throttling configuration. At least it may be good enough for limiting MaxConcurrentCalls.

Actual throttling control is performed by [ChannelHandler](https://github.com/CoreWCF/CoreWCF/blob/main/src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/ChannelHandler.cs). 
My idea was to inject my custom ServiceThrottle object with configured MaxConcurrentCalls to ChannelHandler constructor.
In CoreWCF code ServiceThrottle object (taken from `channel.GetProperty<ServiceThrottle>`) is injected to ChannelHandler by [ServiceDispatcher/CreateServiceChannelDispatcherAsync](https://github.com/CoreWCF/CoreWCF/blob/main/src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/ServiceDispatcher.cs).

I was able to I inject my instance of ServiceThrottle object in `Startup.ConfigureServices()`. 
And result was even promising. I could observe that throttling is partially working.
On every call values in ServiceThrottle were increasing.
But for MaxConcurrentCalls it doesn't work because method `ChannelHandler.RequestInfo.Cleanup` which releases counter is not called at all.

**So it looks that implementation is not finished.**

---

Following code can be used for testing in ServiceThrottle is applied to channel.
```
    public class MyServiceBehavior : IServiceBehavior
    {
        public void AddBindingParameters(ServiceDescription serviceDescription, ServiceHostBase serviceHostBase, Collection<ServiceEndpoint> endpoints, BindingParameterCollection bindingParameters)
        {
        }

        public void ApplyDispatchBehavior(ServiceDescription serviceDescription, ServiceHostBase serviceHostBase)
        {
            foreach (ChannelDispatcher dispatcher in serviceHostBase.ChannelDispatchers)
            {
                // Attempt 1
                dispatcher.ChannelInitializers.Add(new MyChannelInitializer());

                // Attempt 2
                dispatcher.Opening += Dispatcher_Opening;
                dispatcher.Opened += Dispatcher_Opened;

                // Attempt 3
                // dispatcher.DispatchRuntime.InstanceProvider = new MyInstanceProvider();
            }
        }

        private void Dispatcher_Opening(object sender, System.EventArgs e)
        {
            var channelDispatcher = (ChannelDispatcher)sender;
        }

        private void Dispatcher_Opened(object sender, System.EventArgs e)
        {
            var channelDispatcher = (ChannelDispatcher)sender;
        }

        public void Validate(ServiceDescription serviceDescription, ServiceHostBase serviceHostBase)
        {
        }
    }

    public class MyChannelInitializer : IChannelInitializer
    {
        // This is called when new channel from client is created
        public void Initialize(IClientChannel channel)
        {
            // This is null
            var serviceThrottle = channel.GetProperty<ServiceThrottle>();

            // This is null and private
            // ((CoreWCF.Channels.ServiceChannel)channel).InnerChannel.GetProperty<ServiceThrottle>()
        }
    }

   // While handling calls in service method you can also debug this.
   CoreWCF.OperationContext.Current.InternalServiceChannel.GetProperty<CoreWCF.Dispatcher.ServiceThrottle>() is null
```

**Comment by cobalto** [comment]
Hello,

Any new development on this regard?
